Update CONTRIBUTING.md (#2984)
python-poetry/poetry: CONTRIBUTING.md

* **Provide specific examples to demonstrate the steps**..
* **Describe the current behavior** and **explain which behavior you expected to see instead** and why.

### Contributing to documentation

> One of simplest ways to get started contributing to a project is through improving documentation. Poetry is constantly evolving, this means that sometimes our documentation has gaps. You can help by

Reviewer: ```suggestion
One of the simplest ways to get started contributing to a project is through improving documentation. Poetry is constantly evolving, this means that sometimes our documentation has gaps. You can help by
```

> adding missing sections, editing our content so it is more accessible, creating new content (tutorials, FAQs etc).

Reviewer: ```suggestion
adding missing sections, editing the existing content so it is more accessible or creating new content (tutorials, FAQs, etc).
```

> **Note:** A great way to understand Poetry's design and how it all fits together, is to add FAQ entries for commonly
> asked questions. Poetry members usually mark issues with [candidate/faq](https://github.com/python-poetry/poetry/issues?q=is%3Aissue+label%3Acandidate%2Ffaq+) to indicate that the issue either contains a response
> that explains how something works or might benefit from an entry in the FAQ.

> Issues pertaining to documentation are usually marked with the [Documentation](https://github.com/python-poetry/poetry/labels/Documentation) label.

Reviewer: ```suggestion
Issues pertaining to the documentation are usually marked with the [Documentation](https://github.com/python-poetry/poetry/labels/Documentation) label.
```

* Fill in [the required template](https://github.com/python-poetry/poetry/blob/master/.github/PULL_REQUEST_TEMPLATE.md)
* Be sure that your pull request contains tests that cover the changed or added code.
* If your changes warrant a documentation change, the pull request must also update the documentation.

> **Note:** Make sure your branch is [rebased](https://docs.github.com/en/free-pro-team@latest/github/using-git/about-git-rebase) against latest main branch. A maintainer might ask you to ensure the branch is
> up-to-date prior to merging your Pull Request if changes have conflicts.

All pul requests, unless otherwise instructed, need to be first accepted into the main branch (`master`).

### Issue triage

> **Note:** If you have an issue that haven't had any attention, you can ping us `@python-poetry/triage` on the issue. Please, give us reasonable time to get to your issue first, spamming us with messages
> does not help anyone.

> If helping with the triage of reported issue, this section provides some useful information to assist your contribution.

Reviewer: ```suggestion
If you are helping with the triage of reported issues, this section provides some useful information to assist you in your contribution.
```

#### Triage steps

> 1. If `pyproject.toml` is missing or `-vvv` debug logs (with stack trace) is not provide and required, request that the issue author provide it.

Reviewer: ```suggestion
1. If `pyproject.toml` is missing or `-vvv` debug logs (with stack trace) is not provided and required, request that the issue author provides it.
```

> 1. Attempt to reproduce the issue with reported version or request further clarification from the issue author.

Reviewer: ```suggestion
1. Attempt to reproduce the issue with the reported Poetry version or request further clarification from the issue author.
```

> 1. Ensure the issue is not already resolved. You can attempt to reproduce using latest preview release and/or poetry from the main branch.

Reviewer: ```suggestion
1. Ensure the issue is not already resolved. You can attempt to reproduce using the latest preview release and/or poetry from the main branch.
```

> 1. If issue cannot be reproduced,
>    1. clarify with issue author,
>    1. close issue or notify `@python-potry/triage`.
> 1. If issue can be reproduced,
>    1. comment on the issue confirming so
>    1. notify `@python-potry/triage`.
>    1. if possible, identify root cause of the issue.
>    1. if interested, attempt to fix via a pull request.

Reviewer: ```suggestion
1. If the issue cannot be reproduced,
   1. clarify with the issue's author,
   1. close the issue or notify `@python-poetry/triage`.
1. If the issue can be reproduced,
   1. comment on the issue confirming so
   1. notify `@python-poetry/triage`.
   1. if possible, identify the root cause of the issue.
   1. if interested, attempt to fix it via a pull request.
```

Often times you would want to attempt to reproduce issues with multiple versions of `poetry` at the same time. For these use cases, the [pipx project](https://pipxproject.github.io/pipx/) is useful.

You can set your environment up like so.

```sh
pipx install --suffix @1.0.10 'poetry==1.0.10'
pipx install --suffix @1.1.0rc1 'poetry==1.1.0rc1'
pipx install --suffix @master 'poetry @ git+https://github.com/python-poetry/poetry'
```

> **Hint:** Do not forget to update your `poetry@master` installation in sync with upstream.

For `@local` it is recommended that you do something similar to the following as editable installs are not supported for PEP 517 projects.

```sh
# note this will not work for Windows, and we assume you have already run `poetry install`
cd /path/to/python-poetry/poetry
ln -sf $(poetry run which poetry) ~/.local/bin/poetry@local
```

> **Hint:** This mechanism can also be used to test pull requests.

### Git Workflow

All development work is performed against Poetry's main branch (`master`). All changes are expected to be submitted and accepted to this
branch.

#### Release branch

When a release is ready, the following are required before a release is tagged.

> 1. A release branch with prefix `release-`, eg: `release-1.1.0rc1`.
> 1. A pull request from release branch to main branch (`master`) if minor/major release. Otherwise, to bug fix branch (eg: `1.0`).
>    1. The pull request description MUST include change log corresponding to the release (eg: [#2971](https://github.com/python-poetry/poetry/pull/2971)).
>    1. The pull request must contain a commit that updates [CHANGELOG.md](CHANGELOG.md) and bumps project version (eg: [#2971](https://github.com/python-poetry/poetry/pull/2971/commits/824e7b79defca435cf1d765bb633030b71b9a780)).
>    1. The pull request must have a `Release` label specified.
> 
> Once the branch pull-request is ready and approved, a member of `@python-poetry/core` will,
> 
> 1. Tag the branch with the version identifier (eg: `1.1.0rc1`).
> 2. Merge the pull request once the release is created and assets uploaded by CI.

Reviewer: ```suggestion
1. A release branch with the prefix `release-`, eg: `release-1.1.0rc1`.
1. A pull request from the release branch to the main branch (`master`) if it's a minor or major release. Otherwise, to the bug fix branch (eg: `1.0`).
   1. The pull request description MUST include the change log corresponding to the release (eg: [#2971](https://github.com/python-poetry/poetry/pull/2971)).
   1. The pull request must contain a commit that updates [CHANGELOG.md](CHANGELOG.md) and bumps the project version (eg: [#2971](https://github.com/python-poetry/poetry/pull/2971/commits/824e7b79defca435cf1d765bb633030b71b9a780)).
   1. The pull request must have the `Release` label specified.

Once the branch pull-request is ready and approved, a member of `@python-poetry/core` will,

1. Tag the branch with the version identifier (eg: `1.1.0rc1`).
2. Merge the pull request once the release is created and assets are uploaded by the CI.
```

> **Note:** In this case, we prefer a merge commit instead of squash or rebase merge.

#### Bug fix branch

> Once a minor version (eg: `1.1.0`) is release, a new branch for the minor version (eg: `1.1`) is created for bug fix release. Changes identified

Reviewer: ```suggestion
Once a minor version (eg: `1.1.0`) is released, a new branch for the minor version (eg: `1.1`) is created for the bug fix releases. Changes identified
```

   1. The pull request must contain a commit that updates [CHANGELOG.md](CHANGELOG.md) and bumps project version (eg: [#2971](https://github.com/python-poetry/poetry/pull/2971/commits/824e7b79defca435cf1d765bb633030b71b9a780)).
   1. The pull request must have a `Release` label specified.

Once the branch pull-request is ready and approved, a member of `@python-poetry/core` will,

1. Tag the branch with the version identifier (eg: `1.1.0rc1`).
2. Merge the pull request once the release is created and assets uploaded by CI.

> **Note:** In this case, we prefer a merge commit instead of squash or rebase merge.

#### Bug fix branch

Once a minor version (eg: `1.1.0`) is release, a new branch for the minor version (eg: `1.1`) is created for bug fix release. Changes identified
or acknowledged by the Poetry team as requiring a bug fix can be submitted as a pull requests against this branch.

At the time of writing only issues meeting the following criteria may be accepted into a bug fix branch. Trivial fixes may be accepted on a
case-by-case basis.

> 1. Issue breaks core functionality and/or is a critical regression.

Reviewer: ```suggestion
1. The issue breaks a core functionality and/or is a critical regression.
```

> 1. Change set does not introduce new feature or changes existing functionality.

Reviewer: ```suggestion
1. The change set does not introduce a new feature or changes an existing functionality.
```

1. A new minor release is not expected within a reasonable time frame.

> 1. If issue affects next minor/major release, a corresponding fix has been accepted into the main branch.

Reviewer: ```suggestion
1. If the issue affects the next minor/major release, a corresponding fix has been accepted into the main branch.
```
